FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+
+```python
+da.sel(x=..., y=..., lat=..., lon=...)
+```
+
+The easiest solution for this situation would be to raise an error. Alternatively, we could introduce a new parameter to specify how to combine the resulting integer indexers (i.e., union vs intersection), although this could already be achieved by chaining `.sel()` calls or combining `.sel()` with `.merge()` (it may or may not be straightforward).
+
+### 2.6 Using indexes for data alignment
+
+Another main use if indexes is data alignment in various operations. Some considerations regarding alignment and flexible indexes:
+
+- support for alignment should probably be optional for an `XarrayIndex` subclass.
+  - like `pandas.Index`, the index wrapper classes that support it should implement `.equals()`, `.union()` and/or `.intersection()`
+  - support might be partial if that makes sense (outer, inner, left, right, exact...).
+  - index equality might involve more than just the labels: for example a spatial index might be used to check if the coordinate system (CRS) is identical for two sets of coordinates
+  - some indexes might implement inexact alignment, like in [#4489](https://github.com/pydata/xarray/pull/4489) or a `KDTree` index that selects nearest-neighbors within a given tolerance
+  - alignment may be "multi-dimensional", i.e., the `KDTree` example above vs. dimensions aligned independently of each other
+- we need to decide what to do when one dimension has more than one index that supports alignment
+  - we should probably raise unless the user explicitly specify which index to use for the alignment
+- we need to decide what to do when one dimension has one or more index(es) but none support alignment
+  - either we raise or we fail back (silently) to alignment based on dimension size
+- for inexact alignment, the tolerance threshold might be given when building the index and/or when performing the alignment
+- are there cases where we want a specific index to perform alignment and another index to perform selection?
+  - it would be tricky to support that unless we allow multiple indexes per coordinate
+  - alternatively, underlying indexes could be picked internally in a "meta" index for one operation or another, although the risk is to eventually have to deal with an explosion of index wrapper classes with different meta indexes for each combination that we'd like to use.
+
+### 2.7 Using indexes for other purposes
+
+Xarray also provides a number of Dataset/DataArray methods where indexes are used in various ways, e.g.,
+
+- `resample` (`CFTimeIndex` and a `DatetimeIntervalIndex`)
+- `DatetimeAccessor` & `TimedeltaAccessor` properties (`CFTimeIndex` and a `DatetimeIntervalIndex`)
+- `interp` & `interpolate_na`,
+   - with `IntervalIndex`, these become regridding operations. Should we support hooks for these operations?
```

Comment:
I think indexes should optionally support interpolation.

Reply:
That would certainly be useful.